File: Back-END/todo-api/src/test/java/com/todo/api/service/TodoServiceTest.java
package com.todo.api.service;

import com.todo.api.model.Todo;
import com.todo.api.repository.TodoRepository;
import jakarta.persistence.EntityNotFoundException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
@DisplayName("Todo Service Tests")
class TodoServiceTest {

    @Mock
    private TodoRepository todoRepository;

    @InjectMocks
    private TodoService todoService;

    private Todo sampleTodo;

    @BeforeEach
    void setUp() {
        sampleTodo = new Todo();
        sampleTodo.setId(1L);
        sampleTodo.setTitle("Test Todo");
        sampleTodo.setDescription("Test Description");
        sampleTodo.setCompleted(false);
        sampleTodo.setCreatedAt(LocalDateTime.now());
        sampleTodo.setUpdatedAt(LocalDateTime.now());
    }

    @Test
    @DisplayName("Should return all todos")
    void getAllTodos_ShouldReturnAllTodos() {
        // Given
        List<Todo> expectedTodos = Arrays.asList(sampleTodo);
        when(todoRepository.findAll()).thenReturn(expectedTodos);

        // When
        List<Todo> actualTodos = todoService.getAllTodos();

        // Then
        assertEquals(expectedTodos, actualTodos);
        verify(todoRepository).findAll();
    }

    @Test
    @DisplayName("Should return todo by ID when exists")
    void getTodoById_WhenTodoExists_ShouldReturnTodo() {
        // Given
        when(todoRepository.findById(1L)).thenReturn(Optional.of(sampleTodo));

        // When
        Todo actualTodo = todoService.getTodoById(1L);

        // Then
        assertEquals(sampleTodo, actualTodo);
        verify(todoRepository).findById(1L);
    }

    @Test
    @DisplayName("Should throw EntityNotFoundException when todo not found")
    void getTodoById_WhenTodoNotExists_ShouldThrowException() {
        // Given
        when(todoRepository.findById(anyLong())).thenReturn(Optional.empty());

        // When & Then
        EntityNotFoundException exception = assertThrows(
            EntityNotFoundException.class,
            () -> todoService.getTodoById(1L)
        );
        
        assertEquals("Todo not found with id: 1", exception.getMessage());
        verify(todoRepository).findById(1L);
    }

    @Test
    @DisplayName("Should return todos by completion status")
    void getTodosByStatus_ShouldReturnFilteredTodos() {
        // Given
        List<Todo> completedTodos = Arrays.asList(sampleTodo);
        when(todoRepository.findByCompleted(true)).thenReturn(completedTodos);

        // When
        List<Todo> actualTodos = todoService.getTodosByStatus(true);

        // Then
        assertEquals(completedTodos, actualTodos);
        verify(todoRepository).findByCompleted(true);
    }

    @Test
    @DisplayName("Should search todos by title")
    void searchTodos_ShouldReturnMatchingTodos() {
        // Given
        String searchTitle = "Test";
        List<Todo> matchingTodos = Arrays.asList(sampleTodo);
        when(todoRepository.findByTitleContainingIgnoreCase(searchTitle)).thenReturn(matchingTodos);

        // When
        List<Todo> actualTodos = todoService.searchTodos(searchTitle);

        // Then
        assertEquals(matchingTodos, actualTodos);
        verify(todoRepository).findByTitleContainingIgnoreCase(searchTitle);
    }

    @Test
    @DisplayName("Should create new todo")
    void createTodo_ShouldSaveAndReturnTodo() {
        // Given
        Todo newTodo = new Todo();
        newTodo.setTitle("New Todo");
        newTodo.setDescription("New Description");
        
        when(todoRepository.save(any(Todo.class))).thenReturn(sampleTodo);

        // When
        Todo createdTodo = todoService.createTodo(newTodo);

        // Then
        assertEquals(sampleTodo, createdTodo);
        verify(todoRepository).save(newTodo);
    }

    @Test
    @DisplayName("Should update existing todo")
    void updateTodo_WhenTodoExists_ShouldUpdateAndReturnTodo() {
        // Given
        Todo updateDetails = new Todo();
        updateDetails.setTitle("Updated Title");
        updateDetails.setDescription("Updated Description");
        updateDetails.setCompleted(true);
        
        when(todoRepository.findById(1L)).thenReturn(Optional.of(sampleTodo));
        when(todoRepository.save(any(Todo.class))).thenReturn(sampleTodo);

        // When
        Todo updatedTodo = todoService.updateTodo(1L, updateDetails);

        // Then
        assertEquals(sampleTodo, updatedTodo);
        verify(todoRepository).findById(1L);
        verify(todoRepository).save(sampleTodo);
        assertEquals("Updated Title", sampleTodo.getTitle());
        assertEquals("Updated Description", sampleTodo.getDescription());
        assertTrue(sampleTodo.isCompleted());
    }

    @Test
    @DisplayName("Should delete todo when exists")
    void deleteTodo_WhenTodoExists_ShouldDeleteTodo() {
        // Given
        when(todoRepository.findById(1L)).thenReturn(Optional.of(sampleTodo));

        // When
        todoService.deleteTodo(1L);

        // Then
        verify(todoRepository).findById(1L);
        verify(todoRepository).delete(sampleTodo);
    }

    @Test
    @DisplayName("Should toggle todo completion status")
    void toggleTodoStatus_WhenTodoExists_ShouldToggleStatus() {
        // Given
        boolean originalStatus = sampleTodo.isCompleted();
        when(todoRepository.findById(1L)).thenReturn(Optional.of(sampleTodo));
        when(todoRepository.save(any(Todo.class))).thenReturn(sampleTodo);

        // When
        Todo toggledTodo = todoService.toggleTodoStatus(1L);

        // Then
        assertEquals(sampleTodo, toggledTodo);
        assertEquals(!originalStatus, sampleTodo.isCompleted());
        verify(todoRepository).findById(1L);
        verify(todoRepository).save(sampleTodo);
    }
}

File: Front-END/todo-app/src/__tests__/TodoService.test.ts
import axios from 'axios';
import { TodoService } from '../services/TodoService';
import { Todo } from '../types/Todo';

// Mock axios
jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

const mockTodo: Todo = {
  id: 1,
  title: 'Test Todo',
  description: 'Test Description',
  completed: false,
  dueDate: '2025-09-21T15:00:00Z',
  createdAt: '2025-09-14T10:00:00Z',
  updatedAt: '2025-09-14T10:00:00Z'
};

describe('TodoService', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('getAllTodos', () => {
    test('should fetch all todos successfully', async () => {
      const mockTodos = [mockTodo];
      mockedAxios.get.mockResolvedValue({ data: mockTodos });

      const result = await TodoService.getAllTodos();

      expect(mockedAxios.get).toHaveBeenCalledWith('http://localhost:8080/api/todos');
      expect(result).toEqual(mockTodos);
    });

    test('should handle error when fetching todos fails', async () => {
      const errorMessage = 'Network Error';
      mockedAxios.get.mockRejectedValue(new Error(errorMessage));

      await expect(TodoService.getAllTodos()).rejects.toThrow(errorMessage);
    });
  });

  describe('getTodoById', () => {
    test('should fetch todo by id successfully', async () => {
      mockedAxios.get.mockResolvedValue({ data: mockTodo });

      const result = await TodoService.getTodoById(1);

      expect(mockedAxios.get).toHaveBeenCalledWith('http://localhost:8080/api/todos/1');
      expect(result).toEqual(mockTodo);
    });
  });

  describe('getTodosByStatus', () => {
    test('should fetch todos by completion status', async () => {
      const completedTodos = [{ ...mockTodo, completed: true }];
      mockedAxios.get.mockResolvedValue({ data: completedTodos });

      const result = await TodoService.getTodosByStatus(true);

      expect(mockedAxios.get).toHaveBeenCalledWith('http://localhost:8080/api/todos/status?completed=true');
      expect(result).toEqual(completedTodos);
    });
  });

  describe('searchTodos', () => {
    test('should search todos by title', async () => {
      const searchResults = [mockTodo];
      mockedAxios.get.mockResolvedValue({ data: searchResults });

      const result = await TodoService.searchTodos('Test');

      expect(mockedAxios.get).toHaveBeenCalledWith('http://localhost:8080/api/todos/search?title=Test');
      expect(result).toEqual(searchResults);
    });

    test('should encode search query properly', async () => {
      const searchResults = [mockTodo];
      mockedAxios.get.mockResolvedValue({ data: searchResults });

      await TodoService.searchTodos('Test & Query');

      expect(mockedAxios.get).toHaveBeenCalledWith('http://localhost:8080/api/todos/search?title=Test%20%26%20Query');
    });
  });

  describe('createTodo', () => {
    test('should create new todo successfully', async () => {
      const newTodo = { title: 'New Todo', description: 'New Description', completed: false };
      mockedAxios.post.mockResolvedValue({ data: mockTodo });

      const result = await TodoService.createTodo(newTodo);

      expect(mockedAxios.post).toHaveBeenCalledWith('http://localhost:8080/api/todos', newTodo);
      expect(result).toEqual(mockTodo);
    });
  });

  describe('updateTodo', () => {
    test('should update todo successfully', async () => {
      const updatedTodo = { ...mockTodo, title: 'Updated Todo' };
      mockedAxios.put.mockResolvedValue({ data: updatedTodo });

      const result = await TodoService.updateTodo(1, updatedTodo);

      expect(mockedAxios.put).toHaveBeenCalledWith('http://localhost:8080/api/todos/1', updatedTodo);
      expect(result).toEqual(updatedTodo);
    });
  });

  describe('deleteTodo', () => {
    test('should delete todo successfully', async () => {
      mockedAxios.delete.mockResolvedValue({});

      await TodoService.deleteTodo(1);

      expect(mockedAxios.delete).toHaveBeenCalledWith('http://localhost:8080/api/todos/1');
    });
  });

  describe('toggleTodoStatus', () => {
    test('should toggle todo status successfully', async () => {
      const toggledTodo = { ...mockTodo, completed: true };
      mockedAxios.patch.mockResolvedValue({ data: toggledTodo });

      const result = await TodoService.toggleTodoStatus(1);

      expect(mockedAxios.patch).toHaveBeenCalledWith('http://localhost:8080/api/todos/1/toggle');
      expect(result).toEqual(toggledTodo);
    });
  });
});